perfkitbenchmarker/vpn_service.py

```
@@ -0,0 +1,326 @@
+    return key
+
+  def Create(self, gateway_pair, suffix=''):
+    self.GWPair = gateway_pair
+    self.name = self.getKeyFromGWPair(gateway_pair)
+    self.tunnel_config = TunnelConfig(tunnel_name=self.name, suffix=suffix)
+
+  def Delete(self):
+    pass
+
+  def GetVPN(self, gateway_pair, suffix=''):
+    """Gets a VPN object for the gateway_pair or creates one if none exists
+
+    Args:
+    gateway_pair: a tuple of two VPNGWs
+    """
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    if benchmark_spec is None:
+      raise errors.Error('GetVPN called in a thread without a '
+                         'BenchmarkSpec.')
+    with benchmark_spec.vpns_lock:
+      key = self.getKeyFromGWPair(gateway_pair, suffix)
+      if key not in benchmark_spec.vpns:
+        self.Create(gateway_pair, suffix)
+        benchmark_spec.vpns[key] = self
+      return benchmark_spec.vpns[key]
+
+  def ConfigureTunnel(self):
+    """Configure the VPN tunnel."""
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    vpn_gateway_0 = benchmark_spec.vpn_gateways[self.GWPair[0]]
+    vpn_gateway_1 = benchmark_spec.vpn_gateways[self.GWPair[1]]
+
+    assert not (vpn_gateway_0.require_target_to_init and vpn_gateway_1.require_target_to_init), 'Cant connect 2 passive VPN GWs'
+
+    tunnel_config_hash = None
+    while not self.isTunnelConfigured():
+        vpn_gateway_0.ConfigureTunnel(self.tunnel_config)
+        vpn_gateway_1.ConfigureTunnel(self.tunnel_config)
+        if self.tunnel_config.hash() == tunnel_config_hash:
+          raise ValueError('Not enough info to configure tunnel.')
+        tunnel_config_hash = self.tunnel_config.hash()
+
+    tunnel_status = self.isTunnelReady()
+    logging.info('Tunnel is ready?: %s ' % tunnel_status)
+
+  def isTunnelConfigured(self):
+    """Returns True if the tunnel configuration is complete.
+
+    Returns:
+      boolean.
+    """
+    is_tunnel_configured = False
+    if len(self.tunnel_config.endpoints) == 2:
```

If the number of endpoints isn't two should you throw an exception?

---

Not necessarily, we might not have registered them yet. This is the control for the earlier while setup loop, so its expecting things not to be ready immediately.